src/profile-logic/marker-data.js

```diff
@@ -355,23 +362,51 @@ export function correlateIPCMarkers(threads: Thread[]): IPCMarkerCorrelations {
   // message seqno, and message type. Since the seqno is only unique for each
   // message channel pair, we use the PIDs and message type as a way of
   // identifying which channel pair generated this message.
-  function makeMarkerID(srcPid, destPid, data) {
-    return `${srcPid},${destPid},${data.messageSeqno},${data.messageType}`;
+  function makeMarkerID(thread, data) {
+    let pids;
+    if (data.direction === 'sending') {
+      pids = `${thread.pid},${data.otherPid}`;
+    } else {
+      pids = `${data.otherPid},${thread.pid}`;
+    }
+    return pids + `,${data.messageSeqno},${data.messageType}`;
   }
 
   function formatThreadName(tid) {
     const name = threadNames.get(tid);
     if (name === undefined) {
-      throw new Error('unable to find thread name');
+      return undefined;
     }
     return `${name} (Thread ID: ${tid})`;
   }
 
-  // First, construct a mapping of marker IDs to destination markers for
-  // faster lookup. We also collect the friendly thread names while we have
-  // access to all the threads. It's considerably more difficult to do this
-  // processing later.
-  const dstMarkers: Map<string, Object> = new Map();
+  function phaseToIndex(data): number {
+    if (data.direction === 'sending') {
+      switch (data.phase) {
+        case 'transferBegin':
+          return 1;
+        case 'transferEnd':
+          return 2;
+        default:
+          // 'endpoint'
+          return 0;
+      }
+    } else {
+      switch (data.phase) {
+        case 'transferEnd':
+          return 3;
+        default:
+          // 'endpoint'
+          return 4;
+      }
+    }
+  }
+
+  // First, construct a mapping of marker IDs to an array of markers with that
+  // ID for faster lookup. We also collect the friendly thread names while we
+  // have access to all the threads. It's considerably more difficult to do
+  // this processing later.
+  const markersByKey: Map<string, Array<Object>> = new Map();
```

Comment:
Flow's `Object` type is basically an `any` type, so it provides almost no type safety. Please create an inline type for this array's Object. I think it would provide better type safety, and be easier for future readers to maintain,

Reply:
Fixed.